### DISCUSSION TITLE
Problems with the "Bigger Application" tutorial regarding `include_router` + breaking change suggestion

### DISCUSSION THREAD

**Discussion by papb** [fastapi/fastapi#5491] [Questions] [answered]
The [Bigger Applications](https://fastapi.tiangolo.com/tutorial/bigger-applications/) tutorial recommends using several `APIRouter` instances and then using `app.include_router` on them. After doing this and having some problems (which I'll outline below), I am wondering if the behavior should be changed or if the documentation should be improved.

First, some facts after debugging and digging into FastAPI's source code:

- `app.include_router()` is just `app.router.include_router()`
- `router1.include_router(router2)` just [copies over all routes](https://github.com/tiangolo/fastapi/blob/81115dba53d9bb4a5d6433b086f63a24cfd2f7c9/fastapi/routing.py#L708) from `router2` into `router1` at the moment it is called. Therefore, in the end, when there is an incoming request, there is just one single router (`app.router`) to handle it.
- When a request with a trailing slash is handled, the `redirect_slashes` option of that single router is checked.

This way, setting the `redirect_slashes` option of any child router is irrelevant. It currently does not give an error, warning, anything. Rather, the only existent possibility is doing `app.router.redirect_slashes = False` which will impact all routes.

Funnily enough, for my specific case what I wanted to do was to disable redirecting slashes on the whole application, so I managed to do what I wanted simply with `app.router.redirect_slashes = False` and forgetting about the `redirect_slashes` option on child routers. Still I wanted to open this discussion because I think the current behavior is misleading (at best). Maybe you'll consider it a bug.

There is another slight problem in my mind about this. When I first saw `include_router`, the mental model I created about it was that the parent router would "store a pointer to it", so to speak. Therefore, I was expecting that if I added new routes to a child router after doing `app.include_router`, the new routes would be visible from the parent router (but they're not, since `include_router` actually just copies the routes over at the moment it's called).

I suggest the following:

- Breaking change: remove the `redirect_slashes` option from `APIRouter()` and add a `redirect_slashes` option to `FastAPI()`
  - Update [this code](https://github.com/tiangolo/fastapi/blob/81115dba53d9bb4a5d6433b086f63a24cfd2f7c9/fastapi/applications.py#L122-L134) to pass the new `redirect_slashes` option to the root router
- Update the "Bigger Application" tutorial to explain that the routes are copied over at the moment `include_router` is called, so that further adding routes to it will not work

I could create a Pull Request doing the above, but first I'd like to see the opinions of maintainers (and other people).

Thank you! Nice work on FastAPI!

**Comment by slafs** [comment]
Hah. I just run into a different problem that's caused exactly by this issue. Specifically, by the fact that `include_router` copies over the routes from the given (sub) router.

In our app we have a custom "auth check" class that is provided to each endpoint with `Depends`. We wanted some common information (that can be derived from that dependency) to be visible in the endpoint's description and then on the auto-generated API docs (killer feature of FastAPI BTW ❤️).
I couldn't find a better way, so I've created a custom `APIRoute` subclass, which on `__init__` appended some text to `description` argument.

ATM using a custom route class is possible only via (sub)routers, so we create our router like this:
```py
router = APIRouter(route_class=OurCustomRoute)
```
and then later include that router on the "main" app via `include_router` method.

Now, because of the way `include_router` works, **our custom route class gets instantiated two times for each endpoint**, resulting in description being amended twice. First time when the endpoint is "registered" with the router (i.e. when decorating an endpoint with e.g. `@router.get(...)`)
and second time when calling `include_router` (via `add_api_route` method on the router).

If there's a better way to achieve the above, I'm open to suggestions, but nevertheless this behaviour, as @papb already noticed, was surprising and results in some hacky workarounds in our codebase.

**Comment by tiangolo** [comment, marked as answer]
I want to refactor that so that it's indeed a pointer to the child router, that will also allow several other things. But there's a lot of details to handle, corner cases, and background compatibility to handle, so I had to pause that. But hopefully at some point I'll be able to continue on that and figure it out.

**Comment by rcrogers** [comment]
Just lost an hour to order-of-operations mattering way more than a declarative syntax would suggest!
